Add REST light platform

robhowlett opened the pull request:
## Description:

Add support for lights using a REST API.

**Related issue (if applicable):** N/A

**Pull request in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io) with documentation (if applicable):** home-assistant/home-assistant.github.io#3757

## Example entry for `configuration.yaml` (if applicable):
```yaml
light:
  - platform: rest
    resource: http://IP_ADDRESS/ENDPOINT
```

## Checklist:

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io)

If the code communicates with devices, web services, or third-party tools:
  - [x] Local tests with `tox` run successfully. **Your PR cannot be merged unless tests pass**
  - [x] New dependencies have been added to the `REQUIREMENTS` variable ([example][ex-requir]).
  - [x] New dependencies are only imported inside functions that use them ([example][ex-import]).
  - [x] New dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.
  - [x] New files were added to `.coveragerc`.


[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L14
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L54

robhowlett committed:
Add REST light platform

robhowlett committed:
Fix linter errors

TopdRob commented:
Can this also work with GET requests? Specifying different parameters for the ON request and OFF request? I have a light which I can control with these REST webservice requests:
- http://raspberrypi:5001/control?device=light&state=on
- http://raspberrypi:5001/control?device=light&state=off

robhowlett commented:
@TopdRob This platform does not support GET requests to make changes. Within the REST specification, GET requests are used to retrieve data and therefore I don't believe using GET to make a change would be within the specifications of a REST interface. As a result I don't feel that it should be within the scope of this platform.

tinloaf commented on homeassistant/components/light/rest.py:
```diff
@@ -0,0 +1,432 @@
+            async_render_with_possible_json_value(text, self._name)
+        self._name = result
+
+        return req
```
I don't think you need to return `req` here, as you've `yield`ed from it above.

MartinHjelmare replied:
The return value is used in `async_setup_platform`.

robhowlett replied:
@tinloaf - my understanding is the same as @MartinHjelmare - I believe I need to return it as  the return value is used in `async_setup_platform`. Please let me know if I've misunderstood.

tinloaf commented on homeassistant/components/light/rest.py:
```diff
@@ -0,0 +1,432 @@
+        else:
+            self._state = None
+
+        result = self._supported_features_template.\
```
That's interesting, I'm not sure if there is already a platform that may change its `supported_features` at any time. Have you tried whether the front end supports this? Maybe there need changes to be made there to update the list of supported features.

robhowlett replied:
Thanks @tinloaf. I've just checked this and I can confirm that the front end does support it. If the light is turned off and on, then `supported_features` is updated and reflected in the front end.

robhowlett committed:
Change variable name for better clarity

robhowlett committed:
Merge branch 'dev' into rest-light

robhowlett committed:
Merge branch 'dev' into rest-light

robhowlett commented:
@pvizeli - are there any other steps that I need to perform before this PR can be approved? This is my first time going through this process. Thanks.

pvizeli commented:
I will look at it. Sorry for the long delay

robhowlett commented:
Thanks @pvizeli

Danielhiversen commented:
@pvizeli could you have a look?

pvizeli reviewed (changes requested):
rename `CONF_RESOURCE` to `CONF_URL_ON`, `CONF_URL_OFF` and `CONF_URL_STATE`

A rest API should look like:
`curl -X POST http://mybulb/on`
`curl -X POST http://mybulb/off`
`curl http://mybulb/state`

That absolute the BODY_ON/BODY_OFF stuff. That solve the json dump code on turn_on and after that it look good :)

balloob commented:
Any update on this PR?

robhowlett commented:
Apologies for the delay - I'll be making the changes this week.

robhowlett committed:
Allow different endpoints for on/off/state

This is in line with the suggestions made by @pvizeli. The body_on and
body_off parameters have been retained to allow interaction with all
REST lights whether they use parameters or URL to pass on/off (there is
debate online as to the best approach). It also retains compatibility
with how the REST switch component works in Home Assistant.

robhowlett committed:
Fix pylint errors

robhowlett commented:
@pvizeli - I've now made the suggested changes. I've left the `BODY_ON` and `BODY_OFF` config items so that the platform can talk to devices that use parameters rather than the URL to pass information (online debate doesn't seem to show a preference for one or the other when using REST) - this also keeps it in step with how switch.rest platform in Home Assistant works. I hope you're ok with this approach.

balloob commented on homeassistant/components/light/rest.py:
```diff
@@ -0,0 +1,440 @@
+    vol.Optional(CONF_BODY_OFF, default=DEFAULT_BODY_OFF): cv.template,
+    vol.Optional(CONF_BODY_ON, default=DEFAULT_BODY_ON): cv.template,
+    vol.Optional(CONF_BODY_BRIGHTNESS,
+                 default=DEFAULT_BODY_BRIGHTNESS): cv.template,
```
All these defaults don't work anymore. We have recently upgraded voluptuous and all defaults will be passed through the validator. So make all defaults just strings and they will be converted to template.

balloob commented on homeassistant/components/light/rest.py:
```diff
@@ -0,0 +1,440 @@
+
+
+# pylint: disable=unused-argument
+@asyncio.coroutine
```
We've migrated to `async`/`await` syntax. Please drop this decorator and prefix async methods with `async` like `async def async_setup_platform`. Replace any `yield from` with `await`.

balloob commented on homeassistant/components/light/rest.py:
```diff
@@ -0,0 +1,440 @@
+        else:
+            async_add_devices([light])
+    except (TypeError, ValueError):
+        _LOGGER.error("Missing resource or schema in configuration. "
```
I don't think that this is possible if we validate the url ?

balloob commented on homeassistant/components/light/rest.py:
```diff
@@ -0,0 +1,440 @@
+            body_brightness_t = self._body_brightness.async_render() % \
+                kwargs[ATTR_BRIGHTNESS]
+            brightness_dict = json.loads(body_brightness_t)
+            body_on_dict.update(brightness_dict)
```
Oh no. This is not allowed.

We have a hard rule to not use templates to generate JSON. That's a big no no. This should all be removed.

balloob commented on homeassistant/const.py:
```diff
@@ -139,6 +139,9 @@
 CONF_UNIT_OF_MEASUREMENT = 'unit_of_measurement'
 CONF_UNIT_SYSTEM = 'unit_system'
 CONF_URL = 'url'
+CONF_URL_OFF = 'url_off'
```
These should not be added to the main const file, just keep them in the platform.

balloob reviewed (changes requested):
We do not allow using templates to generate other data structures. Templates will generate strings and that's it.

If you want to sent data structures, use a `complex_template` validator and then have the user define the structure in YAML:

```
light:
  platform: rest
  turn_on:
    state: true
    brightness: '{{ brightness }}'
    rgb_color: '{{ rgb_color }}'
```

You can then render by importing the `render_complex` method from `helpers.template` .

robhowlett commented:
Closing PR as other commitments mean I unfortunately don't have time for this currently.
